## Weekly activity
1. _Suggest two ways and write codes to display two images simultaneously_. You can use any image snapped from your handphone,  downloaded from internet or images from week 4 materials on MS teams. The two images are a color image and its corresponding grayscale image. 
2. Write codes that performs the following:
    - Load the video “img_pexels.mp4” into the Python environment, *resize* it and display the videos with smaller frames (The frames can be of any size, as long as it is smaller). You can specify an arbitrary frame rate.
    - Save it as a separate files: “smaller_img_pexels.avi” or "smaller_img_pexels.mp4"
3. Create a **random noise color and grayscale** image. You can set a custom width and height. (Hint: use Numpy functions like `np.array` and `np.reshape`.)
4. Extract the region of interest (flower) from the 'flower.jfif'.
5. Enlarge the image "dog.jfif" by using different techniques: 1) Linear interpolation, 2) Cubic interpolation and 3) nearest neighbor interpolation. Comment on the upscaled of all the outputs.

In [1]:
#import open and numpy to the workplace
import sys
#python 3.7 is required
assert sys.version_info >= (3,7)

import cv2 as cv
import numpy as np

#For reproducibility
np.random.seed(99)

#make sure that optimisation is enabled
if not cv.useOptimized():
    cv.setUseOptimized(True)
    
cv.useOptimized()

True

# Answer for question 1

In [6]:
# Question 1 method 1

skyline_img = cv.imread('skyline.jpeg')
skyline_img_grayscale = cv.imread('skyline.jpeg',0)

cv.imshow("original",skyline_img)
cv.imshow("grayscale",skyline_img_grayscale)

cv.waitKey(0)    
cv.destroyAllWindows()

In [8]:
# Question 1 method 2

skyline_img = cv.imread('skyline.jpeg')
skyline_img_grayscale = cv.cvtColor(skyline_img,cv.COLOR_BGR2GRAY)

cv.imshow("original",skyline_img)
cv.imshow("grayscale",skyline_img_grayscale)


cv.waitKey(0)    
cv.destroyAllWindows()

# Answer for question 2

In [10]:

# Create a VideoCapture object
cap = cv.VideoCapture('img_pexels.mp4')

#define the codec and create VideoWriter object
fourcc = cv.VideoWriter_fourcc(*'XVID')
fps = 20
width = int(cap.get(3))
height = int(cap.get(4))
output = cv.VideoWriter('smaller_img_pexels.avi', fourcc, fps, (width, height))


# Check if the object has been created successfully
if not cap.isOpened():
    print("Unable to create video")

# Read until the video is completed.
while cap.isOpened():
    ret, frame = cap.read()
    
    # if frame is read then ret is True
    if not ret:
        print("Can't receive frame.")
        break
        
    
    output.write(frame)
    
    cv.namedWindow('frame',cv.WINDOW_NORMAL)
    cv.imshow('frame', frame)
    # Press Esc key to exit (27 is ASCII code for Esc). cv.waitKey() returns 32 bit integer values
    if cv.waitKey(1) & 0xFF == 27:
        break

# destroy the constructor
cap.release()
output.release()
cv.destroyAllWindows()

# Answer for question 3

In [24]:
from PIL import Image

array = np.random.random((720,1080))

img = Image.fromarray(array,'L')
img.show()

cv.waitKey(0)    
cv.destroyAllWindows()

# Answer for question 4

In [26]:
#load the flower picture
flower = cv.imread('flower.jfif')

print(flower.shape)

edited_flower = flower[20:125,80:180,0:3]
flower = cv.imshow('original',edited_flower)

cv.waitKey(0)
cv.destroyAllWindows()

(183, 275, 3)


# Answer for question 5

In [2]:
#import open and numpy to the workplace
import sys
#python 3.7 is required
assert sys.version_info >= (3,7)

import cv2 as cv
import numpy as np

#For reproducibility
np.random.seed(99)

#make sure that optimisation is enabled
if not cv.useOptimized():
    cv.setUseOptimized(True)
    
cv.useOptimized()

True

In [3]:
#load the original file
dog_img = cv.imread('dog.jfif')
cv.imshow('original',dog_img)

cv.waitKey(0)
cv.destroyAllWindows()

In [11]:
#Linear Interpolation
dog_img_resize_linear = cv.resize(dog_img, None, fx = 1.5, fy = 1.5, interpolation = cv.INTER_LINEAR)
cv.imshow('Upsize_Linear', dog_img_resize_linear)

cv.waitKey(0)
cv.destroyAllWindows()

121 µs ± 996 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [9]:
#Cubic interpolation
dog_img_resize_cubic = cv.resize(dog_img, None, fx = 1.5, fy = 1.5, interpolation = cv.INTER_CUBIC)
cv.imshow('Upsize_Cubic',dog_img_resize_cubic)

cv.waitKey(0)
cv.destroyAllWindows()

240 µs ± 2.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [12]:
#nerest neightbor interpolation
dog_img_resize_nearest = cv.resize(dog_img, None, fx = 1.5, fy = 1.5, interpolation = cv.INTER_NEAREST)
cv.imshow('Upsize_Nearest',dog_img_resize_nearest)

cv.waitKey(0)
cv.destroyAllWindows()

67.7 µs ± 1.67 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


## Comment on the upscaled of all the outputs

| Criteria             | Liner Interpolation| Cubic Interpolation| Nearest Neighbor Interpolation |
| :---                 |    :----:          |          :----:    |                           ----:|
| Processing time      | medium             | highest            |             lowest             |
| File size            | medium             | largest            |            smallest            |


**Description of the above table**.

1. Processing time **(Nearest < Linear < Cubic)**

Cubic Interpolation has the required the highest processing time followed by Linear Interpoaltion and Nearest Neightbor Interpoaltion. Cubic Interpolation has the most complex computation among the three Interpolation.

<img title="a title" alt="Alt text" src="Processing time.PNG">
***by using %timeit function we are able to show the computational time of each interpolation. Linear required 121 microsecond, cubic required 240 microsecond, nearest required 67.7 microsecond***.


2. File size **(Cubic > Linear > Nearest)**
As aforementioned, Cubic Interpolation has the most complex computation but it also produce the image with highest quality among the other two.

<img title="a title" alt="Alt text" src="File_size.PNG">
***By saving the outputs of all the interpolations, we can find that the file sizes are different by checking the "properties" of the files. In which, Cubic have the largest file size(), Linear is the one in the middle () and Nearest has the smallest file size(). The three files hve the same pixel count, but the Nearest is compressed into fewer bytes, hence the smallest file size. ***

